## ひろしまQuest2020#stayhome【球種予測部門】

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 100)

In [2]:
ALL_MERGE = 'intermediate/all_merge_5.f'

In [3]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 150)

### 不要な列を削除

In [4]:
all_pitch.drop(
    columns=[
        '年度', 
        '試合ID', 
        'ホームチームID', 'アウェイチームID', 
        '投手ID', '投手チームID', 
        '打者ID', '打者チームID', 
        'プレイ前走者状況', 
        '捕手ID', 
        'opening_date', 'game_date',
        'start_time', 'game_time', 'elapsed_time'
    ], inplace=True)

### Rename

In [5]:
all_pitch.rename(columns={
    'データ内連番': 'No',
    '投球位置区域': 'pitch_area',
    '試合内連番': 'No_in_game',
    '試合内投球数': 'pitch_cnt_in_game',
    'イニング': 'inning',
    'イニング内打席数': 'bat_cnt_in_inning',
    '打席内投球数': 'pitch_cnt_in_bat',
    '投手登板順': 'pitch_order',
    '投手試合内対戦打者数': 'player_cnt_in_game',
    '投手試合内投球数': 'pitcher_cnt_in_game',
    '投手イニング内投球数': 'pitcher_cnt_in_inning',
    '打者打順': 'bat_order',
    '打者試合内打席数': 'bat_cnt_in_game',
    'プレイ前ホームチーム得点数': 'home_point',
    'プレイ前アウェイチーム得点数': 'away_point',
    'プレイ前アウト数': 'out_cnt',
    'プレイ前ボール数': 'ball_cnt',
    'プレイ前ストライク数': 'strike_cnt',
}, inplace=True)

In [6]:
print(all_pitch.shape)
all_pitch.head()

(778767, 135)


No  ball  pitch_area  No_in_game  pitch_cnt_in_game  inning  \
0   0   0.0         0.0           1                  1       1   
1   1   0.0         8.0           2                  2       1   
2   2   0.0         5.0           3                  3       1   
3   3   0.0        12.0           4                  4       1   
4   4   0.0         8.0           5                  5       1   

   bat_cnt_in_inning  pitch_cnt_in_bat  pitch_order  player_cnt_in_game  \
0                  1                 1            1                   1   
1                  1                 2            1                   1   
2                  1                 3            1                   1   
3                  2                 1            1                   2   
4                  2                 2            1                   2   

   pitcher_cnt_in_game  pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  \
0                    1                      1          1                1   
1                    2                      2          1                1   
2                    3                      3          1                1   
3                    4                      4          2                1   
4                    5                      5          2                1   

   home_point  away_point  out_cnt  ball_cnt  strike_cnt  first  second  \
0           0           0        0         0           0      0       0   
1           0           0        0         0           1      0       0   
2           0           0        0         0           2      0       0   
3           0           0        1         0           0      0       0   
4           0           0        1         0           1      0       0   

   third  base_cnt  pitch_LR  bat_LR  top_bot  role  pos_pit  \
0      0         0       0.0     1.0        1     1        0   
1      0         0       0.0     1.0        1     1        0   
2      0         0       0.0     1.0        1     1        0   
3      0         0       0.0     0.0        1     1        0   
4      0         0       0.0     0.0        1     1        0   

   date_from_opening  elapsed_min  nighter  ce-pa  league  home  point_diff  \
0                  0     0.000000        1      0       1     1           0   
1                  0     0.200000        1      0       1     1           0   
2                  0     0.383333        1      0       1     1           0   
3                  0     1.016667        1      0       1     1           0   
4                  0     1.200000        1      0       1     1           0   

   runner_23  bant  squize  cleanup  pinch  fourball  savepoint  ball_cnt_0-0  \
0          0     0       0        1      0         0          0             1   
1          0     0       0        1      0         0          0             0   
2          0     0       0        1      0         0          0             0   
3          0     0       0        1      0         0          0             1   
4          0     0       0        1      0         0          0             0   

   ball_cnt_0-1  ball_cnt_0-2  ball_cnt_0-3  ball_cnt_1-0  ball_cnt_1-1  \
0             0             0             0             0             0   
1             0             0             0             1             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             1             0   

   ball_cnt_1-2  ball_cnt_1-3  ...  firm_bat  height_bat  weight_bat  \
0             0             0  ...         0         176          74   
1             0             0  ...         0         176          74   
2             0             0  ...         0         176          74   
3             0             0  ...         0         180          74   
4             0             0  ...         0         180          74   

   draft_o

### train
- 行数: 257117

In [7]:
train = all_pitch.dropna(subset=['pitch_area'])
train.shape

(257117, 135)

### test
- 行数: 521650

In [8]:
test = all_pitch[all_pitch['pitch_area'].isnull()]
test.shape

(521650, 135)

In [9]:
del all_pitch
gc.collect()

60

In [10]:
train_d = train.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

In [11]:
test_d = test.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

## Dataset作成

In [12]:
lgb_train = lgb.Dataset(train_d, train['ball'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [13]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 8,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 6.9923570049658075, 
        'lambda_l2': 0.002378623984798833, 
        'num_leaves': 18, 
        'feature_fraction': 0.45199999999999996, 
        'bagging_fraction': 0.9799724836460725, 
        'bagging_freq': 4, 
        'min_child_samples': 20
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=8000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
print('Best CV score:', cv_results['multi_logloss-mean'][-1])

[100]	cv_agg's multi_logloss: 1.15408 + 0.00198243
[200]	cv_agg's multi_logloss: 1.1323 + 0.00220181
[300]	cv_agg's multi_logloss: 1.12523 + 0.00219871
[400]	cv_agg's multi_logloss: 1.12091 + 0.00218255
[500]	cv_agg's multi_logloss: 1.11782 + 0.00229762
[600]	cv_agg's multi_logloss: 1.11553 + 0.0022767
[700]	cv_agg's multi_logloss: 1.11388 + 0.00241613
[800]	cv_agg's multi_logloss: 1.11274 + 0.00243222
[900]	cv_agg's multi_logloss: 1.112 + 0.00245632
[1000]	cv_agg's multi_logloss: 1.11159 + 0.00250339
[1100]	cv_agg's multi_logloss: 1.11137 + 0.00261299
[1200]	cv_agg's multi_logloss: 1.11134 + 0.00268821
[1300]	cv_agg's multi_logloss: 1.11139 + 0.00276887
Best num_boost_round: 1257
Best CV score: 1.1112402566616455
CPU times: user 1h 27min 29s, sys: 1min 42s, total: 1h 29min 11s
Wall time: 27min 48s


In [14]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [15]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 27min 21s, sys: 26.6 s, total: 27min 47s
Wall time: 8min 18s


### Feature Importance

In [16]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
14                ball_cnt      2290
128            age_dif_b-c      2315
116            age_dif_p-b      2373
60               curve_pit      2477
83              height_bat      2505
84              weight_bat      2521
63                fork_pit      2534
6       player_cnt_in_game      2587
125            bmi_dif_p-c      2612
95         salary_year_bat      2648
118  salary_x_year_dif_p-b      2742
114         salary_dif_p-b      2842
130  salary_x_year_dif_b-c      2863
89        bat_game_cnt_bat      2902
126         salary_dif_b-c      2936
31              point_diff      2948
61              slider_pit      3070
97                 bmi_bat      3090
117    salary_year_dif_p-b      3322
4         pitch_cnt_in_bat      3333
129    salary_year_dif_b-c      3601
88          batter_cnt_bat      3627
131            bmi_dif_b-c      3987
0               No_in_game      4053
1        pitch_cnt_in_game      4173
119            bmi_dif_p-b      4270
8    pitcher_cnt_in_inning      4504
26             elapsed_min      4771
7      pitcher_cnt_in_game      5269
25       date_from_opening      7176

In [17]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 18min 3s, sys: 8.42 s, total: 18min 12s
Wall time: 5min 35s


## 結果出力

In [18]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 9)


index         0         1         2         3         4         5  \
0      0  0.413858  0.029387  0.131606  0.262305  0.002558  0.000017   
1      1  0.353499  0.078649  0.111449  0.385564  0.004143  0.000021   
2      2  0.323407  0.086266  0.139202  0.262360  0.085644  0.000084   
3      3  0.349801  0.095721  0.176498  0.280091  0.018275  0.000039   
4      4  0.228129  0.040685  0.276376  0.341709  0.012589  0.000037   

          6         7  
0  0.000015  0.160253  
1  0.000012  0.066664  
2  0.000030  0.103007  
3  0.000015  0.079561  
4  0.000018  0.100458

### 2017の平均
- ストレート           0.468254
- カーブ             0.074725
- スライダー           0.185806
- シュート            0.070633
- フォーク            0.083013
- チェンジアップ         0.054217
- シンカー            0.011361
- カットボール          0.051992

In [19]:
submit.mean()

index    260824.500000
0             0.521186
1             0.076621
2             0.182393
3             0.055728
4             0.069884
5             0.043342
6             0.006939
7             0.043907
dtype: float64

In [20]:
submit.to_csv('submit/lgbm_5_tune.csv', header=False, index=False)

### 評価結果
#### 5-tune (135)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- CV score: 1.1112402566616455
- 評価結果  : 1.3991780

#### 5 (135)
- 'num_leaves' : 31,
- 'learning_rate' : 0.1,
- CV score: 1.1161825304353852
- 評価結果  : 1.4252286

#### 4 (128)
- CV score: 1.1158111693845418
- 評価結果  : 1.4299265

#### 3 (97)
- CV score: 1.1172290313076036
- 評価結果  : 1.4335806